SECURITY.WHITEPAPER.md

```diff
@@ -0,0 +1,129 @@
+# Datasafe Security Whitepaper
+
+## General information
+Datasafe is a flexible encryption library. It uses different encryption algorithms. Some of them can be 
+configured by client application. Under the hood Datasafe uses BouncyCastle library to perform encryption.
+
+CMS (Cryptographic Message Syntax) standard [RFC5652](https://tools.ietf.org/html/rfc5652.html) employed for storing private 
+files encrypted with symmetric keys as well as for sharing files with other users using asymmetric key pairs. 
+
+## Default locations
+By default on system start have to be configured one system dfs (distributed file system).
+It is used for storing users' private profile, public profile, keystore with public keys, keystore with private keys.
+
+Default location within system dfs:
+```
+    /profiles
+        /private
+            /username - user's private profile 
+        /public
+            /username - user's public profile   
+    /users
+        /public
+            /pubkeys - keystore consists user's public key
+            /inbox - location of shared with user files    
+        /private
+            /keystore - keystore consists user's private key
+            /files/SIV - location of private files. SIV is 3 symbol path encryption algorithm identifier.
+```                   
+Example private profile:
+```
+    {
+        "keystore": {
+            "resource": "s3://bucketname/users/username/private/keystore"
+        },
+        "privateStorage": [
+            [{"id": "DEFAULT"}, {"resource": "s3://bucketname/users/username/private/files/"}]
+        ],
+        "inboxWithFullAccess": {
+            "resource": "s3://bucketname/users/username/public/inbox/"
+        },
+        "publishPublicKeysTo": {
+            "resource": "s3://bucketname/users/username/public/pubkeys"
+        },
+        "associatedResources": [
+            {"resource": "s3://bucketname/users/username/"}
+        ],
+        "documentVersionStorage": {
+            "resource": "s3://bucketname/users/username/versions/"
+        },
+        "appVersion": "BASELINE"
+    }
+```
+            
+Example public profile:
+```
+    {
+        "publicKeys": {
+            "resource": "s3://bucketname/users/username/public/pubkeys"
+        },
+        "inbox": {
+            "resource": "s3://bucketname/users/username/public/inbox/"
+        },
+        "appVersion": "BASELINE"
+    }   
+```
+    
+## Keystore encryption
+System wide password is used to open keystore and users' personal password to read users' keys. Password can be changed 
+without the need of changing keystore content.
+By default used:
+-  BCFKS keystore type to store on disk and UBER to cache keys in memory;
```

Comment:
Please provide links to keystore security and algorithms used for UBER. Add note why it is ok to use TripleDES for in-memory keys - because if someone gets to memory dump of machine there is really not that much can be done to protect application and it has good performance